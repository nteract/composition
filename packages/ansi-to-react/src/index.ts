import * as React from "react";
import { ansiToJson, AnserJsonEntry } from "anser";
import { escapeCarriageReturn } from "escape-carriage";

const LINK_REGEX = /(https?:\/\/(?:www\.|(?!www))[^\s\.]+\.[^\s]{2,}|www\.[^\s]+\.[^\s]{2,})/;

/**
 * Converts ANSI strings into JSON output.
 * @name ansiToJSON
 * @function
 * @param {String} input The input string.
 * @return {Array} The parsed input.
 */
function ansiToJSON(input: string) {
  input = escapeCarriageReturn(input);
  return ansiToJson(input, {
    json: true,
    remove_empty: true
  });
}

/**
 * Converts an Anser bundle into a React Node.
 * @param linkify whether links should be converting into clickable anchor tags.
 * @param bundle Anser output.
 */
function convertBundleIntoReact(
  linkify: boolean,
  bundle: AnserJsonEntry,
  key: number
) {
  const style: { backgroundColor?: string; color?: string } = {};
  if (bundle.bg) {
    style.backgroundColor = `rgb(${bundle.bg})`;
  }
  if (bundle.fg) {
    style.color = `rgb(${bundle.fg})`;
  }

  if (!linkify) {
    return React.createElement("span", { style, key }, bundle.content);
  }

  const words = bundle.content.split(" ").reduce(
    (words: React.ReactNode[], word: string, index: number) => {
      // If this isn't the first word, re-add the space removed from split.
      if (index !== 0) {
        words.push(" ");
      }

      // If  this isn't a link, just return the word as-is.
      if (!LINK_REGEX.test(word)) {
        words.push(word);
        return words;
      }

      words.push(
        React.createElement(
          "a",
          {
            key: index,
            href: word,
            target: "_blank"
          },
          `${word}`
        )
      );
      return words;
    },
    [] as React.ReactNode[]
  );
  return React.createElement("span", { style, key }, words);
}

declare type Props = {
  children: string;
  className?: string;
  linkify?: boolean;
};

export default function Ansi(props: Props) {
  return React.createElement(
    "code",
    { className: props.className },
    ansiToJSON(props.children).map(
      convertBundleIntoReact.bind(null, props.linkify)
    )
  );
}
