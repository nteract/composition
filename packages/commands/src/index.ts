import { COMMANDS } from "./commands";
import { makeCommandSetsComponent } from "./components";
import { CellTarget, CommandLocation, GapTarget, makeCellInfo, makeCellModelFromState, makeGapInfo, makeGapModel } from "./model";

export { COMMANDS };
export { CellTarget, CommandLocation, GapTarget, makeCellInfo, makeCellModelFromState, makeGapInfo, makeGapModel };

export const CellCommands =
  makeCommandSetsComponent(makeCellInfo, makeCellModelFromState, COMMANDS.CELL);

export const GapCommands =
  makeCommandSetsComponent(makeGapInfo, makeGapModel, COMMANDS.GAP);
