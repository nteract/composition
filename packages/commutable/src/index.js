/* @flow */

import type { Notebook as v4Notebook } from './v4';

import type {
  ImmutableNotebook,
  JSONType,
} from './types';

type PlaceholderNotebook = {
  nbformat: number,
  nbformat_minor: number,
};

const v4 = require('./v4');

const {
  emptyNotebook,
  emptyCodeCell,
  emptyMarkdownCell,
  appendCell,
  monocellNotebook,
} = require('./structures');

export type Notebook = PlaceholderNotebook & v4Notebook;

function freezeReviver(k: string, v: JSONType): JSONType {
  return Object.freeze(v);
}

// Expected usage of below is
// fromJS(parseNotebook(string|buffer))

function parseNotebook(notebookString: string): Notebook {
  return JSON.parse(notebookString, freezeReviver);
}

function fromJS(notebookJSON: Notebook): ImmutableNotebook {
  if (notebookJSON.nbformat === 4 && notebookJSON.nbformat_minor >= 0) {
    if (Array.isArray(notebookJSON.cells) && typeof notebookJSON.metadata === 'object') {
      return v4.fromJS(notebookJSON);
    }
  }

  if (notebookJSON.nbformat) {
    throw new TypeError(
      `nbformat v${notebookJSON.nbformat}.${notebookJSON.nbformat_minor} not recognized`
    );
  }

  throw new TypeError('This notebook format is not supported');
}

function toJS(immnb: ImmutableNotebook): v4Notebook {
  if (immnb.get('nbformat') === 4 && immnb.get('nbformat_minor') >= 0) {
    return v4.toJS(immnb);
  }
  throw new TypeError('Only notebook format 4 is supported');
}

// Expected usage is stringifyNotebook(toJS(immutableNotebook))
function stringifyNotebook(notebook: v4Notebook): string {
  return JSON.stringify(notebook, null, 2);
}

module.exports = {
  emptyCodeCell,
  emptyMarkdownCell,
  emptyNotebook,
  appendCell,
  monocellNotebook,
  toJS,
  fromJS,
  stringifyNotebook,
};

export type {
  ImmutableNotebook,
};
