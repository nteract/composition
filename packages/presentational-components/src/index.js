// @flow

import { Input } from "./components/input.js";
import { Outputs } from "./components/outputs.js";
import { Pagers } from "./components/pagers.js";
import { Prompt, PromptBuffer } from "./components/prompt.js";
import { Source } from "./components/source.js";
import { Cell } from "./components/cell.js";
import { Cells } from "./components/cells.js";
import { HeaderEditor } from "./components/header-editor.js";
import * as themes from "./themes";

export * from "./styles";

export {
  themes,
  Input,
  Outputs,
  Pagers,
  Prompt,
  PromptBuffer,
  Source,
  Cell,
  Cells,
  HeaderEditor
};
