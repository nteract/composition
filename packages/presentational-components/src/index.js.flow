type CellProps = {
    /**
     * Indicates if a cell is selected
     */
    isSelected: boolean,
    /**
     * Indicates if hovering over a cell
     */
    _hovered: boolean,
    /**
     * Style children when a cell is selected or hovered over
     */
    children?: React.Node
  };
  type CellsProps = {
    children: React.ChildrenArray<any>
  };

  export type AuthorObject = {
    name: string
  };
  export type HeaderDataProps = {
    authors: Array<AuthorObject>,
    title: string,
    description: string,
    tags: Array<string>
  };
  
  export type HeaderEditorProps = {
    /**
     * The data that the header should be populated with.
     */
    headerData: HeaderDataProps,
    /**
     * Whether or not the fields of the header can be edited.
     */
    editable: boolean,
    /**
     * An event handler to run whenever header fields are modified.
     */
    onChange: HeaderDataProps => void,
    /**
     * The theme of the header.
     */
    theme: "light" | "dark"
  };
  
  export type HeaderEditorState = {
    editMode: "none" | "author" | "tag"
  };
  export type InputProps = {
    children: React.Node,
    /**
     * Whether or not to render the children.
     */
    hidden: boolean
  };
  export type OutputsProps = {
    children: React.Node,
    /**
     * Whether or not to collapse the children of the component.
     */
    expanded: boolean,
    /**
     * Whether or not to show the children of the component.
     */
    hidden: boolean
  };

  export type PagersProps = {
    children: React.Node,
    hidden: boolean
  };
  export type SourceProps = {
    language: string,
    children: React.Node,
    className?: string,
    theme: "light" | "dark"
  }; 
  type HighlighterProps = {
    language: string,
    className: string,
    children: React.Node,
    theme: "light" | "dark"
  };
