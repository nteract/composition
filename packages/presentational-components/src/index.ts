
import { Input } from "./components/input";
import { Outputs } from "./components/outputs";
import { Pagers } from "./components/pagers";
import { Prompt, PromptBuffer } from "./components/prompt";
import { Source } from "./components/source";
import { Cell } from "./components/cell";
import { Cells } from "./components/cells";
import { HeaderEditor } from "./components/header-editor";
import * as themes from "./themes";

export * from "./styles";

export {
  themes,
  Input,
  Outputs,
  Pagers,
  Prompt,
  PromptBuffer,
  Source,
  Cell,
  Cells,
  HeaderEditor
};
