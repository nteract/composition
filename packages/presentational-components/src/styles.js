// @flow

import * as React from "react";
import styles from "@nteract/styles";

export class Styles extends React.Component<*> {
  render() {
    return (
      <React.Fragment>
        <style dangerouslySetInnerHTML={{ __html: styles }} />
        {this.props.children}
      </React.Fragment>
    );
  }
}
