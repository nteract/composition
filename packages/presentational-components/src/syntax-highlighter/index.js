// @flow
import SyntaxHighlighter from "react-syntax-highlighter/prism";
import * as React from "react";
import {
  xonokai as darkTheme,
  vs as lightTheme
} from "react-syntax-highlighter/styles/prism";

// import syntax from "./theme";

type HighlighterProps = {
  language: string,
  className: string,
  children: React.Node,
  theme: "light" | "dark"
};

const Highlighter = (props: HighlighterProps) => {
  let language = props.language;
  if (language === "ipython") {
    language = "python";
  } else if (language === "text/x-scala") {
    // Adjust for the scala codemirror type
    language = "scala";
  } else if (language.startsWith("text/x-")) {
    // Strip off the language from the mimetype
    language = language.slice("text/x-".length);
  }

  return (
    <SyntaxHighlighter
      style={props.theme === "light" ? lightTheme : darkTheme}
      language={language}
      className={props.className}
      customStyle={{
        padding: "10px 0px 10px 10px",
        margin: "0px",
        backgroundColor: "var(--cm-background, #fafafa)",
        border: "none"
      }}
    >
      {props.children}
    </SyntaxHighlighter>
  );
};

Highlighter.defaultProps = {
  theme: "light",
  language: "text",
  children: "",
  className: "input"
};

export default Highlighter;
