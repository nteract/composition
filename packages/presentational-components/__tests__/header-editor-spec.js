import * as React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";

import { HeaderEditor } from "../src/";

describe("Header Editor", () => {
  it("renders correctly with no props", () => {
    const tree = renderer.create(<HeaderEditor />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders a static view when not editable", () => {
    const tree = renderer.create(<HeaderEditor editable={false} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly given header data", () => {
    const onChange = jest.fn();

    const headerData = {
      authors: [{ name: "zombo.com" }, { name: "internets" }],
      title: "Welcome to ZomboCom",
      description: `This ... is ... ZomboCom. Welcome. This is ZomboCom; welcome ... to ZomboCom. You can do anything at ZomboCom. Anything at all. The only limit is yourself. Welcome ... to ZomboCom.
Welcome ... to ZomboCom. This is ... ZomboCom. Welcome ... to ZomboCom! This is ZomboCom, welcome! Yes ... This ... is ZomboCom.
This is ZomboCom! And welcome to you, who have come to ZomboCom. Anything ... is possible ... at ZomboCom. You can do ... anything at ZomboCom. The infinite is possible at ZomboCom. The unattainable is unknown at ZomboCom. Welcome to ZomboCom. This ... is ZomboCom.
Welcome to ZomboCom. Welcome. This ... is ... ZomboCom. Welcome ... to ZomboCom! Welcome ... to ZomboCom.`,
      tags: ["best of web", "colorful", "pulsating discs"]
    };

    const wrapper = shallow(
      <HeaderEditor headerData={headerData} onChange={onChange} editable />
    );

    // Click to add an author
    wrapper.find(".author-button").simulate("click");

    const authorEntry = wrapper.find(".author-entry");
    authorEntry.simulate("change", { target: { value: "the world" } });
    // JSX Style and enzyme are not playing well -- it's turning arrays into objects
    // expect(wrapper).toMatchSnapshot();
    authorEntry.simulate("keypress", { key: "Enter" });

    /*
    Can't seem to make all this work...
    expect(onChange).toHaveBeenCalledWith({
      a: "b"
    });*/
  });
});
