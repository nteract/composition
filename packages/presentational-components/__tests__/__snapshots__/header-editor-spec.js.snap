// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Editor renders a static view when not editable 1`] = `
<header
  className="jsx-2279537447"
>
  <div
    className="jsx-2279537447"
    style={
      Object {
        "background": "#EEE",
        "padding": "10px",
      }
    }
  >
    <h1
      className="bp3-heading"
    >
      <div
        className="bp3-editable-text bp3-disabled bp3-editable-text-placeholder"
        onFocus={[Function]}
        tabIndex={null}
      >
        <span
          className="bp3-editable-text-content"
          style={
            Object {
              "height": 0,
              "lineHeight": "0px",
              "minWidth": 80,
            }
          }
        >
          Edit title...
        </span>
      </div>
    </h1>
    <div
      className="jsx-2279537447"
    >
      <span
        className="bp3-popover-wrapper"
      >
        <span
          className="bp3-popover-target"
          onBlur={[Function]}
          onFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <button
            className="bp3-button bp3-disabled bp3-minimal author-button"
            disabled={true}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            tabIndex={-1}
            type="button"
          >
            <span
              className="bp3-icon bp3-icon-add"
              icon="add"
            >
              <svg
                data-icon="add"
                height={16}
                viewBox="0 0 16 16"
                width={16}
              >
                <desc>
                  add
                </desc>
                <path
                  d="M10.99 6.99h-2v-2c0-.55-.45-1-1-1s-1 .45-1 1v2h-2c-.55 0-1 .45-1 1s.45 1 1 1h2v2c0 .55.45 1 1 1s1-.45 1-1v-2h2c.55 0 1-.45 1-1s-.45-1-1-1zm-3-7c-4.42 0-8 3.58-8 8s3.58 8 8 8 8-3.58 8-8-3.58-8-8-8zm0 14c-3.31 0-6-2.69-6-6s2.69-6 6-6 6 2.69 6 6-2.68 6-6 6z"
                  fillRule="evenodd"
                />
              </svg>
            </span>
          </button>
        </span>
      </span>
    </div>
    <div
      className="jsx-2279537447"
    >
      <span
        className="bp3-popover-wrapper"
      >
        <span
          className="bp3-popover-target"
          onBlur={[Function]}
          onFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <button
            className="bp3-button bp3-disabled bp3-minimal"
            disabled={true}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            tabIndex={-1}
            type="button"
          >
            <span
              className="bp3-icon bp3-icon-add"
              icon="add"
            >
              <svg
                data-icon="add"
                height={16}
                viewBox="0 0 16 16"
                width={16}
              >
                <desc>
                  add
                </desc>
                <path
                  d="M10.99 6.99h-2v-2c0-.55-.45-1-1-1s-1 .45-1 1v2h-2c-.55 0-1 .45-1 1s.45 1 1 1h2v2c0 .55.45 1 1 1s1-.45 1-1v-2h2c.55 0 1-.45 1-1s-.45-1-1-1zm-3-7c-4.42 0-8 3.58-8 8s3.58 8 8 8 8-3.58 8-8-3.58-8-8-8zm0 14c-3.31 0-6-2.69-6-6s2.69-6 6-6 6 2.69 6 6-2.68 6-6 6z"
                  fillRule="evenodd"
                />
              </svg>
            </span>
          </button>
        </span>
      </span>
    </div>
    <div
      className="jsx-2279537447"
      style={
        Object {
          "marginTop": "10px",
        }
      }
    >
      <div
        className="bp3-editable-text bp3-disabled bp3-editable-text-placeholder bp3-multiline"
        onFocus={[Function]}
        tabIndex={null}
      >
        <span
          className="bp3-editable-text-content"
          style={
            Object {
              "height": 0,
            }
          }
        >
          Edit description...
        </span>
      </div>
    </div>
  </div>
</header>
`;

exports[`Header Editor renders correctly with no props 1`] = `
<header
  className="jsx-2279537447"
>
  <div
    className="jsx-2279537447"
    style={
      Object {
        "background": "#EEE",
        "padding": "10px",
      }
    }
  >
    <h1
      className="bp3-heading"
    >
      <div
        className="bp3-editable-text bp3-editable-text-placeholder"
        onFocus={[Function]}
        tabIndex={0}
      >
        <span
          className="bp3-editable-text-content"
          style={
            Object {
              "height": 0,
              "lineHeight": "0px",
              "minWidth": 80,
            }
          }
        >
          Edit title...
        </span>
      </div>
    </h1>
    <div
      className="jsx-2279537447"
    >
      <span
        className="bp3-popover-wrapper"
      >
        <span
          className="bp3-popover-target"
          onBlur={[Function]}
          onFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <button
            className="bp3-button bp3-minimal author-button"
            onClick={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="bp3-icon bp3-icon-add"
              icon="add"
            >
              <svg
                data-icon="add"
                height={16}
                viewBox="0 0 16 16"
                width={16}
              >
                <desc>
                  add
                </desc>
                <path
                  d="M10.99 6.99h-2v-2c0-.55-.45-1-1-1s-1 .45-1 1v2h-2c-.55 0-1 .45-1 1s.45 1 1 1h2v2c0 .55.45 1 1 1s1-.45 1-1v-2h2c.55 0 1-.45 1-1s-.45-1-1-1zm-3-7c-4.42 0-8 3.58-8 8s3.58 8 8 8 8-3.58 8-8-3.58-8-8-8zm0 14c-3.31 0-6-2.69-6-6s2.69-6 6-6 6 2.69 6 6-2.68 6-6 6z"
                  fillRule="evenodd"
                />
              </svg>
            </span>
          </button>
        </span>
      </span>
    </div>
    <div
      className="jsx-2279537447"
    >
      <span
        className="bp3-popover-wrapper"
      >
        <span
          className="bp3-popover-target"
          onBlur={[Function]}
          onFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <button
            className="bp3-button bp3-minimal"
            onClick={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="bp3-icon bp3-icon-add"
              icon="add"
            >
              <svg
                data-icon="add"
                height={16}
                viewBox="0 0 16 16"
                width={16}
              >
                <desc>
                  add
                </desc>
                <path
                  d="M10.99 6.99h-2v-2c0-.55-.45-1-1-1s-1 .45-1 1v2h-2c-.55 0-1 .45-1 1s.45 1 1 1h2v2c0 .55.45 1 1 1s1-.45 1-1v-2h2c.55 0 1-.45 1-1s-.45-1-1-1zm-3-7c-4.42 0-8 3.58-8 8s3.58 8 8 8 8-3.58 8-8-3.58-8-8-8zm0 14c-3.31 0-6-2.69-6-6s2.69-6 6-6 6 2.69 6 6-2.68 6-6 6z"
                  fillRule="evenodd"
                />
              </svg>
            </span>
          </button>
        </span>
      </span>
    </div>
    <div
      className="jsx-2279537447"
      style={
        Object {
          "marginTop": "10px",
        }
      }
    >
      <div
        className="bp3-editable-text bp3-editable-text-placeholder bp3-multiline"
        onFocus={[Function]}
        tabIndex={0}
      >
        <span
          className="bp3-editable-text-content"
          style={
            Object {
              "height": 0,
            }
          }
        >
          Edit description...
        </span>
      </div>
    </div>
  </div>
</header>
`;
